### ISSUE TITLE
aurelia-framework and pal not listed as dependency but imported in main.ts/js

### ISSUE THREAD

**Issue by michaelw85** [aurelia/cli#1040]
**I'm submitting a bug report**

* **Library Version:**
Latest (Local aurelia-cli v1.0.0-beta.12)

**Please tell us about your environment:**
* **Operating System:**
Windows [10]

* **Node Version:**
10.15.1

* **NPM Version:**
6.4.1

* **Browser:**
N/A

* **Language:**
all

* **Loader/bundler:**
N/A

**Current behavior:**
aurelia-framework and aurelia-pal are imported in main.ts 
aurelia-pal is imported in main.js

These dependencies are not listed in the package.json.

* **What is the expected behavior?**
Expect dependencies to be listed in the package json.

* **What is the motivation / use case for changing the behavior?**
I think it's not failing because other packages depend on the same packages this seems wrong to me.

**Comment by huochunpeng** [comment]
This is a feature of npm since v3, it is called flat node_modules.

aurelia-framework and aurelia-pal and many other aurelia packages were depa of aurelia-bootstrapper.

That means aurelia requires npm v3+.

Aurelia is not the only thing in JS world now implicitly depending on flat node_modules structure. This is a common practice to reduce the complexity of app’s package.json by hiding deps behind a master dep.

Yarn also provides flat node_modules.

Pnpm is the only package manager that sticks to the old way (npm < v3) by default. But it also provides shamefullyFlatten option (yeah, tell me about the name).

In short, this is not a bug, it is a **feature**.

**Comment by michaelw85** [comment]
Ok I will close this issue if this is done by design.
My typescript linter (suggested microsoft ruleset) disagrees with this setup though, deps should be listed.

**Comment by huochunpeng** [comment]
There is something with your ts linter. Ts linter respects the flat structure.

Most aurelia core team members write apps in TS, they haven’t complained the linter issue.

**Comment by huochunpeng** [comment]
I guess our default linter config does not have the strict rule that you borrowed from ms.

**Comment by huochunpeng** [comment]
I can understand vnext turned that on. It got no massive dep tree :-)

You should check out pnpm, it got the same mindset, that's why they say flat is shameful. With default pnpm, you have to add those missing deps in package.json, otherwise the app would not build.

**Comment by michaelw85** [comment]
I'm running the recommended microsoft ruleset and will try to stick to it as closely as possible. I will be linting on build within a PR flow so when the rule fails the code will not go to master. Thanks for the info though!

**Comment by huochunpeng** [comment]
If you don't need karma (it doesn't play well with pnpm), use pnpm, it forces you to follow the rule, and pnpm is dame dame fast.
